grade-scripts/grades-from-trackers.js

```diff
@@ -0,0 +1,127 @@
+const listManager = require('./shared/list-manager')
+const program = require('commander')
+const fs = require('fs')
+const execSync = require('child_process').execSync
+const chalk = require('chalk')
+
+const Grade = require('../src/classes/grade')
+const https = require('../src/https')
+const trackers = require('../src/trackers')
+const surrogates = require('../src/surrogates')
+const privacyPolicy = require('../src/privacy-policy')
+const scriptUtils = require('./shared/utils')
+
+program
+    .option('-i, --input <name>', 'The name to use when looking for sites, e.g. "test" will look in "test-sites"')
+    .option('-o, --output <name>', 'Output name, e.g. "test" will output files at "test-grades"')
+    .option('-f, --file <name>', 'Allow processing a subset of dumped site data, defined in a file')
+    .parse(process.argv)
+
+const input = program.input
+const inputPath = `${input}-trackers`
+const output = program.output
+const outputPath = `${output}-grades`
+const fileForSubset = program.file
+
+const run = async () => {
+    // load any lists and plug them into any classes that wait for them
+    await listManager.loadLists()
+    trackers.addLists({
+        entityList: listManager.getList('entityList'),
+        whitelist: listManager.getList('whitelist')
+    })
+    surrogates.addLists(listManager.getList('surrogates'))
+    https.addLists({
+        https: listManager.getList('https'),
+        httpsAutoUpgrade: listManager.getList('httpsAutoUpgrade')
+    })
+    privacyPolicy.addLists({
+        tosdr: require('../data/generated/tosdr'),
+        polisis: require('../data/generated/polisis')
+    })
+
+    execSync(`mkdir -p ${outputPath}`)
+
+    // get initial file data
+    let siteDataArray = scriptUtils.getSiteData(inputPath, fileForSubset)
+
+    for (let siteData of siteDataArray) {
+        let url = siteData.url
+        let hostname = url.replace(/https?:\/\//, '')
+
+        if (scriptUtils.dataFileExists(hostname, outputPath)) continue
+
+        siteData.https = https.canUpgradeHost(hostname)
+        siteData.httpsAutoUpgrade = https.hostAutoUpgrades(hostname)
+        siteData.privacyScore = privacyPolicy.getScoreForUrl(url)
+
+        let grade = new Grade(siteData)
+
+        grade.calculate()
+
+        let gradeData = grade.getGrades()
+
+        siteData.score = gradeData
+        siteData.privacy = privacyPolicy.getReasonsForUrl(url)
+
+        console.log(chalk.green(`got grade for ${hostname}: before ${gradeData.site.grade}, after ${gradeData.enhanced.grade}`))
+
+        fs.writeFileSync(`${outputPath}/${hostname}.json`, JSON.stringify(siteData))
```

Comment:
async writeFile might speed this up

Reply:
Not sure, we need to wait for these files to be written before calculating the histogram. Plus, the main blocker is actually HTTPS lists loading rather than file I/O. And we're about to resolve that soon :)